In [1]:
import pandas as pd
import numpy  as np
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings("ignore")

In [2]:
customers = pd.read_csv(r"C:\Analyst Portfolio\E-Commerce-Sales-Analysis\Dataset\olist_customers_dataset.csv")
orders = pd.read_csv(r"C:\Analyst Portfolio\E-Commerce-Sales-Analysis\Dataset\olist_orders_dataset.csv")
orders_items = pd.read_csv(r"C:\Analyst Portfolio\E-Commerce-Sales-Analysis\Dataset\olist_order_items_dataset.csv")
products = pd.read_csv(r"C:\Analyst Portfolio\E-Commerce-Sales-Analysis\Dataset\olist_products_dataset.csv")
payments = pd.read_csv(r"C:\Analyst Portfolio\E-Commerce-Sales-Analysis\Dataset\olist_order_payments_dataset.csv")
reviews = pd.read_csv(r"C:\Analyst Portfolio\E-Commerce-Sales-Analysis\Dataset\olist_order_reviews_dataset.csv")
sellers = pd.read_csv(r"C:\Analyst Portfolio\E-Commerce-Sales-Analysis\Dataset\olist_sellers_dataset.csv")
category = pd.read_csv(r"C:\Analyst Portfolio\E-Commerce-Sales-Analysis\Dataset\product_category_name_translation.csv")
location = pd.read_csv(r"C:\Analyst Portfolio\E-Commerce-Sales-Analysis\Dataset\olist_geolocation_dataset.csv")

In [3]:
#Check Dataset Loaded Successfully
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [4]:
# Check Shape of all data set
print(customers.shape)
print(orders.shape)
print(orders_items.shape)
print(products.shape)
print(payments.shape)
print(reviews.shape)
print(sellers.shape)
print(location.shape)

(99441, 5)
(99441, 8)
(112650, 7)
(32951, 9)
(103886, 5)
(99224, 7)
(3095, 4)
(1000163, 5)


In [5]:
# Check Columns
customers.columns

Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='object')

In [6]:
# Check Columns
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')

In [7]:
# Check Columns
orders_items.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='object')

In [8]:
# Check Columns
products.columns

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='object')

In [9]:
# Check Columns
payments.columns

Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='object')

In [10]:
# Check Columns
reviews.columns

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='object')

In [11]:
# Check Columns
sellers.columns

Index(['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state'], dtype='object')

In [12]:
# Check Columns
location.columns

Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='object')

In [1]:
print(customers.info())
print(orders.info())
print(orders_items.info())
print(products.info())
print(payments.info())
print(reviews.info())
print(sellers.info())
print(location.info())

NameError: name 'customers' is not defined

In [14]:
print(customers.isnull().sum())
print(orders.isnull().sum())
print(orders_items.isnull().sum())
print(products.isnull().sum())
print(payments.isnull().sum())
print(reviews.isnull().sum())
print(sellers.isnull().sum)
print(location.isnull().sum())

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64
product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64
order_id     

In [15]:
print(customers.duplicated().sum())
print(orders.duplicated().sum())
print(orders_items.duplicated().sum())
print(products.duplicated().sum())
print(payments.duplicated().sum())
print(reviews.duplicated().sum())
print(sellers.duplicated().sum)
print(location.duplicated().sum())

0
0
0
0
0
0
<bound method Series.sum of 0       False
1       False
2       False
3       False
4       False
        ...  
3090    False
3091    False
3092    False
3093    False
3094    False
Length: 3095, dtype: bool>
261831


In [16]:
orders.describe(include="all")

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
count,99441,99441,99441,99441,99281,97658,96476,99441
unique,99441,99441,8,98875,90733,81018,95664,459
top,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2018-04-11 10:48:14,2018-02-27 04:31:10,2018-05-09 15:48:00,2018-05-08 23:38:46,2017-12-20 00:00:00
freq,1,1,96478,3,9,47,3,522


In [17]:
payments.describe(include="all")

,order_id,payment_sequential,payment_type,payment_installments,payment_value
count,103886,103886.000000,103886,103886.000000,103886.000000
unique,99440,NaN,5,NaN,NaN
top,fa65dad1b0e818e3ccc5cb0e39231352,NaN,credit_card,NaN,NaN
freq,29,NaN,76795,NaN,NaN
mean,NaN,1.092679,NaN,2.853349,154.100380
std,NaN,0.706584,NaN,2.687051,217.494064
min,NaN,1.000000,NaN,0.000000,0.000000
25%,NaN,1.000000,NaN,1.000000,56.790000
50%,NaN,1.000000,NaN,1.000000,100.000000
75%,NaN,1.000000,NaN,4.000000,171.837500


# Dataset Summary

Customers Table

- Customer ID
- Customer Unique ID
- City
- State

Orders Table

- Order Status
- Purchase Timestamp
- Delivery Timestamp

Order Items

- Product ID
- Seller ID
- Freight Value

Products

- Category
- Weight
- Length
- Height

Payments

- Payment Type
- Installments
- Payment Value

Reviews

- Review Score

Sellers

- Seller City
- Seller State

#  Data Cleaning & Merging


## Check Missing Values

### Cheking Missing  value  for customers

In [18]:
customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

## cheking  Missing Value for Orders

In [19]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [20]:
round(orders.isnull().sum()/len(orders)*100,2)

order_id                         0.00
customer_id                      0.00
order_status                     0.00
order_purchase_timestamp         0.00
order_approved_at                0.16
order_delivered_carrier_date     1.79
order_delivered_customer_date    2.98
order_estimated_delivery_date    0.00
dtype: float64

In [21]:
#Check Order Status
orders.loc[
    orders["order_approved_at"].isnull(),
    "order_status"
].value_counts()

order_status
canceled     141
delivered     14
created        5
Name: count, dtype: int64

In [22]:
#Remove Only Delivered Orders
orders = orders.drop(
    orders[
        (orders["order_status"] == "delivered") &
        (orders["order_approved_at"].isnull())
    ].index
)

In [23]:
#Verify
orders.loc[
    orders["order_approved_at"].isnull(),
    "order_status"
].value_counts()

order_status
canceled    141
created       5
Name: count, dtype: int64

In [24]:
orders["order_approved_at"].isnull().sum()

np.int64(146)

### Handling Missing Values in `order_approved_at`

The `order_approved_at` column contained **160 missing values**.

To understand the reason for these missing values, the corresponding `order_status` values were examined.

The analysis showed:

- **141** canceled orders
- **14** delivered orders
- **5** created orders

Missing approval timestamps for **canceled** and **created** orders were retained because they represent valid business scenarios where an order may not have been approved.

However, the **14 delivered orders** with missing approval timestamps were considered inconsistent, as delivered orders are expected to have an approval timestamp. Therefore, these records were removed from the dataset to maintain data quality and consistency.

## cheking Missing  Value  for orders_Items

In [25]:
orders_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

### Cheking  Missing Values for  products

In [26]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

### Handling Missing Values in `product_category_name`

The `product_category_name` column contained **610 missing values**.

Since these products still exist but their categories are unavailable, removing them would result in unnecessary data loss.

**Action Taken:** Missing values were replaced with **"Unknown"** to preserve all product records while clearly identifying products with unavailable category information.

In [27]:
#filling product_category_name with unknown
products["product_category_name"] = products["product_category_name"].fillna("Unknown")

#### product_name_lenght

The `product_name_lenght` column was left unchanged because it is not required for sales analysis and the missing values do not affect business insights.

In [28]:
products["product_name_lenght"].isnull().sum()

np.int64(610)

The `product_description_lenght` column was retained with missing values since it is not used in the analysis and does not impact sales or profitability calculations.

#### product_photos_qty
Missing values in `product_photos_qty` were replaced with **0**, indicating that photo information was unavailable for those products.

In [29]:
products["product_photos_qty"] = products["product_photos_qty"].fillna(0)

#### Weight & Dimensions (2 missing)

The product dimension columns (`product_weight_g`, `product_length_cm`, `product_height_cm`, and `product_width_cm`) contained only **2 missing records**.

Since these represented a negligible proportion of the dataset, the corresponding rows were removed to maintain data quality.

In [30]:
products = products.dropna(
    subset=[
        "product_weight_g",
        "product_length_cm",
        "product_height_cm",
        "product_width_cm"
    ]
)

In [31]:
products.isnull().sum()

product_id                      0
product_category_name           0
product_name_lenght           609
product_description_lenght    609
product_photos_qty              0
product_weight_g                0
product_length_cm               0
product_height_cm               0
product_width_cm                0
dtype: int64

## Cheking Missing values for  payments 

In [32]:
payments.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

## Cheking Missing values for review

In [33]:
reviews.isnull().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [34]:
reviews["review_comment_title"] = reviews["review_comment_title"].fillna("No Title")

In [35]:
reviews["review_comment_message"] = reviews["review_comment_message"].fillna("No Review")

In [36]:
reviews.isnull().sum()

review_id                  0
order_id                   0
review_score               0
review_comment_title       0
review_comment_message     0
review_creation_date       0
review_answer_timestamp    0
dtype: int64

### Handling Missing Values in Customer Review Columns

The `review_comment_title` and `review_comment_message` columns contained a large number of missing values.

These missing values indicate that many customers provided only a numerical review score without leaving a written title or comment. Since this represents normal customer behavior rather than missing data caused by an error, the review records were retained.

To maintain consistency and simplify text analysis, missing values were replaced with:

- **"No Title"** for `review_comment_title`
- **"No Review"** for `review_comment_message`

This approach preserves all review records while clearly indicating the absence of textual feedback.

## Cheking Missing values for sellers

In [37]:
sellers.isnull().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

## Cheking Missing values for category

In [38]:
category.isnull().sum()

product_category_name            0
product_category_name_english    0
dtype: int64

## Cheking Missing values for location

In [39]:
location.isnull().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

# Check Duplicate Records

In [40]:
customers.duplicated().sum()

np.int64(0)

In [41]:
orders.duplicated().sum()

np.int64(0)

In [42]:
orders_items.duplicated().sum()

np.int64(0)

In [43]:
payments.duplicated().sum()

np.int64(0)

In [44]:
products.duplicated().sum()

np.int64(0)

In [45]:
reviews.duplicated().sum()

np.int64(0)

In [46]:
sellers.duplicated().sum()

np.int64(0)

In [47]:
category.duplicated().sum()

np.int64(0)

In [48]:
location.duplicated().sum()

np.int64(261831)

### Checking Duplicate Records

Duplicate records were examined across all datasets using the `duplicated()` function.

The analysis confirmed that **no duplicate records** were present in any of the datasets.

**Result:** No duplicate records were found, so no rows were removed.

##  Convert Date Columns

The date columns were converted from **object** to **datetime** format to enable accurate date-based analysis, feature engineering, and delivery time calculations.

In [49]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [50]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99427 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99427 non-null  object
 1   customer_id                    99427 non-null  object
 2   order_status                   99427 non-null  object
 3   order_purchase_timestamp       99427 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97644 non-null  object
 6   order_delivered_customer_date  96462 non-null  object
 7   order_estimated_delivery_date  99427 non-null  object
dtypes: object(8)
memory usage: 6.8+ MB


In [51]:
# Convert Date Columns

orders['order_purchase_timestamp'] = pd.to_datetime(
    orders['order_purchase_timestamp']
)

orders['order_approved_at'] = pd.to_datetime(
    orders['order_approved_at']
)

orders['order_delivered_carrier_date'] = pd.to_datetime(
    orders['order_delivered_carrier_date']
)

orders['order_delivered_customer_date'] = pd.to_datetime(
    orders['order_delivered_customer_date']
)

orders['order_estimated_delivery_date'] = pd.to_datetime(
    orders['order_estimated_delivery_date']
)

In [52]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99427 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99427 non-null  object        
 1   customer_id                    99427 non-null  object        
 2   order_status                   99427 non-null  object        
 3   order_purchase_timestamp       99427 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97644 non-null  datetime64[ns]
 6   order_delivered_customer_date  96462 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99427 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.8+ MB


In [53]:
orders_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [54]:
products.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32949 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32949 non-null  object 
 1   product_category_name       32949 non-null  object 
 2   product_name_lenght         32340 non-null  float64
 3   product_description_lenght  32340 non-null  float64
 4   product_photos_qty          32949 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.5+ MB


In [55]:
payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [56]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     99224 non-null  object
 4   review_comment_message   99224 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [57]:
reviews['review_creation_date'] = pd.to_datetime(
    reviews['review_creation_date']
)

reviews['review_answer_timestamp'] = pd.to_datetime(
    reviews['review_answer_timestamp']
)

In [58]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  object        
 1   order_id                 99224 non-null  object        
 2   review_score             99224 non-null  int64         
 3   review_comment_title     99224 non-null  object        
 4   review_comment_message   99224 non-null  object        
 5   review_creation_date     99224 non-null  datetime64[ns]
 6   review_answer_timestamp  99224 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(4)
memory usage: 5.3+ MB


In [59]:
sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [60]:
category.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


## Feature Engineering

Create useful columns.

In [61]:
#Delivery Days
orders["delivery_days"] = (

    orders['order_delivered_customer_date']
    - orders['order_purchase_timestamp']
).dt.days

In [62]:
#Purchase Year
orders['purchase_year'] = (
    orders['order_purchase_timestamp'].dt.year
)

In [63]:
#purchase month
orders["purchase_month"] = (
    orders['order_purchase_timestamp'].dt.month_name()
)

In [64]:
#Purchase Quarter
orders["purchase_quarter"] = (
    orders["order_purchase_timestamp"].dt.quarter
)

In [65]:
#Purchase Weekday
orders["purchase_weekday"] = (
    orders['order_purchase_timestamp'].dt.day_name()
)

In [ ]:
orders.columns

## Translate Product Categories

In [68]:
## Merge English category names.
products = products.merge(
    category,
    on = 'product_category_name',
    how = 'left' 
)

In [69]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm,product_category_name_english_x,product_category_name_english_y
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0,perfumery,perfumery
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0,art,art
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0,sports_leisure,sports_leisure
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0,baby,baby
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0,housewares,housewares


In [70]:
# Merge Orders + Customers
final_dataset = orders.merge(
    customers,
    on='customer_id',
    how='left'
) 

In [71]:
# Merge Order Items
final_dataset = final_dataset.merge(
    orders_items,
    on='order_id',
    how='left'
)

In [72]:
#Merge Products
final_dataset = final_dataset.merge(
    products,
    on='product_id',
    how='left'
)

In [73]:
#Merge Payments
final_dataset = final_dataset.merge(
    payments,
    on ='order_id',
    how ='left'
)

In [74]:
# Merge Reviews
#Before merging, remove duplicate reviews (some orders have multiple review records).
reviews = reviews.drop_duplicates(subset='order_id')

In [75]:
# Then merge:
final_dataset = final_dataset.merge(
    reviews,
    on='order_id',
    how='left'
)

In [76]:
# Merge Sellers
final_dataset = final_dataset.merge(
    sellers,
    on='seller_id',
    how='left'
)

In [77]:
final_dataset.head(5)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,delivery_days,purchase_year,...,payment_value,review_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp,seller_zip_code_prefix,seller_city,seller_state
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,8.0,2017,...,18.12,a54f0611adc9ed256b57ede6b6eb5114,4.0,No Title,"Não testei o produto ainda, mas ele veio corre...",2017-10-11,2017-10-12 03:43:48,9350.0,maua,SP
1,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,8.0,2017,...,2.00,a54f0611adc9ed256b57ede6b6eb5114,4.0,No Title,"Não testei o produto ainda, mas ele veio corre...",2017-10-11,2017-10-12 03:43:48,9350.0,maua,SP
2,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,8.0,2017,...,18.59,a54f0611adc9ed256b57ede6b6eb5114,4.0,No Title,"Não testei o produto ainda, mas ele veio corre...",2017-10-11,2017-10-12 03:43:48,9350.0,maua,SP
3,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,13.0,2018,...,141.46,8d5266042046a06655c8db133d120ba5,4.0,Muito boa a loja,Muito bom o produto.,2018-08-08,2018-08-08 18:37:50,31570.0,belo horizonte,SP
4,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,9.0,2018,...,179.12,e73b67b67587f7644d5bd1a52deb1b01,5.0,No Title,No Review,2018-08-18,2018-08-22 19:07:58,14840.0,guariba,SP


In [78]:
final_dataset.shape

(118419, 46)

In [79]:
final_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118419 entries, 0 to 118418
Data columns (total 46 columns):
 #   Column                           Non-Null Count   Dtype         
---  ------                           --------------   -----         
 0   order_id                         118419 non-null  object        
 1   customer_id                      118419 non-null  object        
 2   order_status                     118419 non-null  object        
 3   order_purchase_timestamp         118419 non-null  datetime64[ns]
 4   order_approved_at                118258 non-null  datetime64[ns]
 5   order_delivered_carrier_date     116345 non-null  datetime64[ns]
 6   order_delivered_customer_date    115022 non-null  datetime64[ns]
 7   order_estimated_delivery_date    118419 non-null  datetime64[ns]
 8   delivery_days                    115022 non-null  float64       
 9   purchase_year                    118419 non-null  int32         
 10  purchase_month                   118419 non-

## Save the Clean Dataset

Create a new folder named Cleaned_Data.

In [80]:
final_dataset.to_csv(
    "ecommerce_master_dataset.csv",
    index=False
)

##  Final Data Validation

The cleaned datasets were successfully merged into a single master dataset.

- Dataset Shape: **118,419 rows × 45 columns**
- Data types were verified and converted where required.
- Feature engineering columns were successfully created.
- Remaining missing values represent valid business scenarios (e.g., canceled orders, missing reviews, or unavailable product information) and were retained to preserve data integrity.

# 📊: Exploratory Data Analysis (EDA)

Exploratory Data Analysis is performed to understand sales performance, customer behavior, product trends, payment methods, delivery performance, and customer satisfaction.

The analysis uses statistical summaries and visualizations to identify important patterns and generate actionable business insights.

In [82]:
total_customers = final_dataset["customer_unique_id"].nunique()
total_orders = final_dataset["order_id"].nunique()
total_products = final_dataset["product_id"].nunique()
total_sellers = final_dataset["seller_id"].nunique()


print("Total Orders:", total_orders)
print("Total Customers:", total_customers)
print("Total Products:", total_products)
print("Total Sellers:", total_sellers)

Total Orders: 99427
Total Customers: 96083
Total Products: 32948
Total Sellers: 3095


## Install Python libraries for  connecting a postsql

In [83]:
!pip install sqlalchemy psycopg2-binary

In [84]:
import pandas as pd
from sqlalchemy import create_engine

In [89]:
from urllib.parse import quote_plus

username = "postgres"
password = quote_plus("Roshan@0310")
host = "localhost"
port = "5432"
database = "ecommerce_analysis"

engine = create_engine(
    f"postgresql+psycopg2://{username}:{password}@{host}:{port}/{database}"
)

In [90]:
final_dataset.to_sql(
    "ecommerce_sales",
    engine,
    if_exists="replace",
    index=False
)

559